components/builder-depot/src/server.rs

```diff
@@ -2307,6 +2264,393 @@ fn target_from_headers(user_agent_header: &UserAgent) -> result::Result<PackageT
     }
 }
 
+fn check_circular_deps(
+    ident: &OriginPackageIdent,
+    target: &PackageTarget,
+    archive: &mut PackageArchive,
+) -> result::Result<(), Response> {
+    let mut pcr_req = JobGraphPackagePreCreate::new();
+    pcr_req.set_ident(format!("{}", ident));
+    pcr_req.set_target(target.to_string());
+
+    let mut pcr_deps = protobuf::RepeatedField::new();
+    let deps_from_artifact = match archive.deps() {
+        Ok(deps) => deps,
+        Err(e) => {
+            info!("Could not get deps from {:#?}: {:#?}", archive, e);
+            return Err(Response::with((status::UnprocessableEntity, "ds:up:4")));
+        }
+    };
+
+    for ident in deps_from_artifact {
+        let dep_str = format!("{}", ident);
+        pcr_deps.push(dep_str);
+    }
+    pcr_req.set_deps(pcr_deps);
+
+    // We need to do it this way instead of via route_message because this function can't be passed
+    // a Request struct, because it's sometimes called from a background thread, and Request
```

Comment:
Since the upstream already checks for circular deps, I don't believe we need to do that check in the downstream (it shouldn't be possible to introduce a circular dependency by downloading a _new_ package from the upstream, since by definition nothing else is pointing to the package in the downstream).

Reply:
👍